shared/storage/minio.py

```diff
@@ -1,133 +1,158 @@
-                    408,
-                    429,
-                    500,
-                    502,
-                    503,
-                    504,
-                ],  # https://cloud.google.com/storage/docs/retry-strategy#python
-            ),
-        )
-        if iam_auth:
-            return Minio(
-                host,
-                secure=verify_ssl,
-                region=region,
-                credentials=ChainedProvider(
-                    providers=[
-                        IamAwsProvider(custom_endpoint=iam_endpoint),
-                        EnvMinioProvider(),
-                        EnvAWSProvider(),
-                    ]
-                ),
-                http_client=http_client,
-            )
+    http_client = urllib3.PoolManager(
+        timeout=Timeout(connect=CONNECT_TIMEOUT, read=READ_TIMEOUT),
+        maxsize=10,
+        cert_reqs="CERT_REQUIRED",
+        ca_certs=os.environ.get("SSL_CERT_FILE") or certifi.where(),
+        retries=Retry(
+            total=5,
+            backoff_factor=1,
+            status_forcelist=[
+                408,
+                429,
+                500,
+                502,
+                503,
+                504,
+            ],  # https://cloud.google.com/storage/docs/retry-strategy#python
+        ),
+    )
+    if iam_auth:
         return Minio(
             host,
-            access_key=access_key,
-            secret_key=secret_key,
             secure=verify_ssl,
             region=region,
+            credentials=ChainedProvider(
+                providers=[
+                    IamAwsProvider(custom_endpoint=iam_endpoint),
+                    EnvMinioProvider(),
+                    EnvAWSProvider(),
+                ]
+            ),
             http_client=http_client,
         )
+    return Minio(
+        host,
+        access_key=access_key,
+        secret_key=secret_key,
+        secure=verify_ssl,
+        region=region,
+        http_client=http_client,
+    )
+
+
+@lru_cache(maxsize=None)
+def get_cached_minio_client(
+    host: str = "",
+    port: str | None = None,
+    access_key_id: str | None = None,
+    secret_access_key: str | None = None,
+    verify_ssl: bool = False,
+    iam_auth: bool = False,
+    iam_endpoint: str | None = None,
+    region: str | None = None,
+):
+    return init_minio_client(
+        host,
+        port,
+        access_key_id,
+        secret_access_key,
+        verify_ssl,
+        iam_auth,
+        iam_endpoint,
+        region,
+    )
+
+
+# Service class for interfacing with codecov's underlying storage layer, minio
+class MinioStorageService(BaseStorageService, PresignedURLService):
+    def __init__(
+        self,
+        minio_config,
+        new_read: bool = False,
+        new_write: bool = False,
+    ):
+        self.minio_config = minio_config
+        self.new_read = new_read
+        self.new_write = new_write
+        self.zstd_default = zstd_decoded_by_default()
```

Comment:
this should probably be moved to a global, as checking with `importlib` every time this is created is probably a very bad idea.